numba_dpex/experimental/launcher.py

```diff
@@ -416,74 +289,71 @@ def _submit_kernel(
+    kernel_sig = types.void(*ty_kernel_args_tuple)
+    # ty_kernel_fn is type specific to exact function, so we can get function
+    # directly from type and compile it. Thats why we don't need to get it in
+    # codegen
+    kernel_dispatcher: KernelDispatcher = ty_kernel_fn.dispatcher
+    kernel_dispatcher.compile(kernel_sig)
+    kernel_module: _KernelModule = kernel_dispatcher.get_overload_device_ir(
         kernel_sig
     )
-    kernel_targetctx = kernel_fn.dispatcher.targetctx
+    kernel_targetctx = kernel_dispatcher.targetctx
 
-    # TODO: refactor so there are no too many locals
-    def codegen(cgctx, builder, sig, llargs):  # pylint: disable=R0914
-        kernel_argtys = kernel_sig.args
-        kernel_args_unpacked = []
-        for pos in range(len(kernel_args)):
-            kernel_args_unpacked.append(builder.extract_value(llargs[2], pos))
+    def codegen(cgctx, builder, sig, llargs):
+        # llargs[0] is kernel function that we don't need anymore (see above)
+        ty_index_space = sig.args[1]
+        ll_index_space = llargs[1]
+        ty_kernel_args_tuple = sig.args[2]
+        ll_kernel_args_tuple = llargs[2]
 
-        fn_body_gen = _LaunchTrampolineFunctionBodyGenerator(
+        generator = _LaunchKernelIRGenerator(
             codegen_targetctx=cgctx,
-            kernel_targetctx=kernel_targetctx,
             builder=builder,
         )
 
-        kernel_bc_byte_str = fn_body_gen.insert_kernel_bitcode_as_byte_str(
-            kernel_module
+        kernel_args = generator.extract_arguments_from_tuple(
+            ty_kernel_args_tuple=ty_kernel_args_tuple,
+            ll_kernel_args_tuple=ll_kernel_args_tuple,
         )
 
-        populated_kernel_args = (
-            fn_body_gen.populate_kernel_args_and_argsty_arrays(
-                kernel_argtys, kernel_args_unpacked
-            )
+        # queue_ref is just a pointer to the attribute, so we don't have to
+        # clean it up
+        queue_ref = generator.get_queue_ref_val(
+            kernel_argtys=ty_kernel_args_tuple,
+            kernel_args=kernel_args,
         )
 
-        qref = fn_body_gen.get_queue_ref_val(
-            kernel_argtys=kernel_args_list,
-            kernel_args=kernel_args_unpacked,
-        )
+        # creates new object, so we must clean it up
+        kernel_ref = generator.get_kernel(queue_ref, kernel_module)
 
-        kbref = fn_body_gen.create_kernel_bundle_from_spirv(
-            queue_ref=qref,
-            kernel_bc=kernel_bc_byte_str,
-            kernel_bc_size_in_bytes=len(kernel_module.kernel_bitcode),
+        # What is it for?
+        (
+            global_range_extents,
+            local_range_extents,
+        ) = generator.create_llvm_values_for_index_space(
+            ty_indexer_arg=ty_index_space,
+            index_arg=ll_index_space,
         )
 
-        kref = fn_body_gen.get_kernel(kernel_module, kbref)
-
-        index_space_values = fn_body_gen.create_llvm_values_for_index_space(
```

Comment:
If you are changing namedtuple to a tuple to get the IDE to better statically analyze the code, better add a type annotation here.